### PR TITLE
fix concurrent map writes error when a large number of connection com…

### LINKED ISSUE #1225: crash when a large number of connections come concurrently
https://github.com/pingcap/tidb/blob/master/perfschema/statement.go#L163

this line will cause `concurrent map writes` error when a large number of connections come concurrently!

### PR DESCRIPTION
**Pull request by liugangnhm**
fix #1225!

PTAL @shenli @coocood

### CODE REVIEW

**Comment by shenli** [comment]
@liugangnhm Please sign the CLA.

**Comment by coocood** [comment]
LGTM

File: perfschema/statement.go [hunk 1]
```diff
@@ -160,7 +160,10 @@ func (ps *perfSchema) StartStatement(sql string, connID uint64, callerName EnumC
 			return nil
 		}
 		source = fmt.Sprintf("%s:%d", fileName, fileLine)
+
+		callerLock.Lock()
```
Comment (shenli) [inline perfschema/statement.go]: Should we lock it before "source, ok = callerNames[callerName]"?

Comment (coocood) [inline perfschema/statement.go]: Or lock before 154 and unlock after 154.

Comment (liugangnhm) [inline perfschema/statement.go]: Yes you are right, we should lock map even if we read it in golang.

 lock is ok because write is only once.

Comment (liugangnhm) [inline perfschema/statement.go]: I had second thoughts again.

- there is only one write op for the same key, so we could read map safely.
- lock out side of `if` could cause performance issue when large number of request comes concurrently.

So I think there is no need to change the code.

Comment (shenli) [inline perfschema/statement.go]: Consider the following two scenarios:
1. Two sessions check an empty callerNames with the same callerName at the same time. They will both write callerNames[callerName] to some source.
2. One session is writing to callerNames[callerName]. In the meanwhile, another session is reading from it.

Maybe scenario 2 will cause race problem. Use a ReadWriteLock will not cause performance issue.

**Comment by shenli** [comment]
LGTM

**Comment by coocood** [comment]
LGTM

**Pull request merged**